lib/iris/tests/unit/cube/test_Cube.py

```diff
@@ -1223,21 +1223,192 @@ def test__lazy(self):
-        self.assertTrue(cube.has_lazy_data())
+    def test_lazy_data(self):
+        for dtype in self.dtypes:
+            data = np.array([0, 1], dtype=dtype)
+            cube = Cube(as_lazy_data(data))
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_with_dtype(self):
+        for dtype in self.dtypes:
+            data = np.array([0, 1], dtype=dtype)
+            cube = Cube(as_lazy_data(data), dtype=dtype)
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_masked__mask_unset(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            cube = Cube(as_lazy_data(data))
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_masked__mask_set(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            data[0] = ma.masked
+            cube = Cube(as_lazy_data(data))
+            self.assertEqual(cube.dtype, np.dtype('float'))
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_masked_with_dtype__mask_unset(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            cube = Cube(as_lazy_data(data), dtype=dtype)
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_masked_with_dtype__mask_set(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            data[0] = ma.masked
+            cube = Cube(as_lazy_data(data), dtype=dtype)
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_real_data_with_dtype_different(self):
+        for dtype in self.dtypes:
+            data = np.array([0, 1], dtype=dtype)
+            emsg = 'Cube does not have lazy data, cannot set dtype'
+            with self.assertRaisesRegexp(ValueError, emsg):
+                Cube(data, dtype=np.complex)
+
+    def test_lazy_data_with_dtype_non_integral(self):
+        for dtype in self.dtypes:
+            data = np.array([0, 1], dtype=dtype)
+            emsg = 'Can only cast lazy data to integral or bool dtype'
```

Comment:
You'll need to change this error message too if you agree with my other suggestions.